### PR TITLE
Add Map Countdown

### PR DESCRIPTION
**Pull request by Back-Slash-N**
Add ability to have a countdown begin at the end of map instead of transitioning instantly.

### CODE REVIEW

File: src/ModFiles/scripts/vscripts/multiplayermod/config.nut [hunk 1]
```diff
@@ -40,6 +40,18 @@ Config_DisableMelCutscenes <- false
 Config_RandomTurret <- false
 Config_RandomTurretLoop <- false
 
+
+// Set true/false if you want the server to wait a bit to allow other players to reach the end of the level before allowing someone to start the elevator (default: true)
+Config_UseCountdown <- true
+
+//! Note: Countdown configs are useless if Config_UseCountdown is set to false.
+// Set to the PERCENTAGE of players that you want to reach the end before the transition enables.
+Config_CountdownPercentage <- 100
```
Comment (OrsellGit) [inline src/ModFiles/scripts/vscripts/multiplayermod/config.nut]: Why a percentage instead of a player count?

File: src/ModFiles/scripts/vscripts/multiplayermod/config.nut [hunk 2]
```diff
@@ -101,7 +113,10 @@ Config_HostOnlyChatCommands <- false
 // ]
 
 Admins <- [
-    // Add Admins Here
+    "[6]303762676", // Qquash
```
Comment (OrsellGit) [inline src/ModFiles/scripts/vscripts/multiplayermod/config.nut]: I don't think you meant to do this

File: src/ModFiles/scripts/vscripts/multiplayermod/hooks.nut [hunk 3]
```diff
@@ -66,6 +66,22 @@ function InstantRun() {
         EntFireByHandle(Entities.FindByName(null, "p2mm_env_global03"), "turnon", "", delay, null, null)
         EntFireByHandle(Entities.FindByName(null, "p2mm_env_global04"), "turnon", "", delay, null, null)
     }
+    if (GetMapName().len() == 37) {
```
Comment (OrsellGit) [inline src/ModFiles/scripts/vscripts/multiplayermod/hooks.nut]: You could instead check if "workshop/" exists in the map name. That means it's a workshop map. Then, you can use a try-catch to make the countdown work. Actually, wait, why did you add this feature for workshop maps? There is no tradition most of the time.

File: src/ModFiles/scripts/vscripts/multiplayermod/hooks.nut [hunk 4]
```diff
@@ -386,6 +402,62 @@ function P2MMLoop() {
                 }
             }
         }
+        
+        // Coop Countdown logic
+        if (Config_UseCountdown && GetMapName().find("mp_coop_lobby_") == null) {
```
Comment (OrsellGit) [inline src/ModFiles/scripts/vscripts/multiplayermod/hooks.nut:391]: Is this just a case for the hub? Maybe you can have this in the map supports.

File: src/ModFiles/scripts/vscripts/multiplayermod/config.nut (hunk 1 above)
Comment (Back-Slash-N) [reply src/ModFiles/scripts/vscripts/multiplayermod/config.nut]: seems like a bad idea, not everyones sessions should have the same set player limit
(i set it to 100 for testing, could probably make that 75% or lower)

File: src/ModFiles/scripts/vscripts/multiplayermod/mapsupport/portal2/sp_a2_bts1.nut [hunk 5]
```diff
@@ -77,8 +80,9 @@ function MapSupport(MSInstantRun, MSLoop, MSPostPlayerSpawn, MSPostMapSpawn, MSO
 
         // Make our own changelevel trigger
         for (local p = null; p = Entities.FindByClassnameWithin(p, "player", Vector(1253.4089355469, -1319.4395751953, -320.2317199707), 104.79999542236);) {
-            EntFire("instanceauto91-exit_fade", "fade")
-            EntFire("p2mm_servercommand", "command", "changelevel sp_a2_bts2", 2)
+            // EntFire("instanceauto91-exit_fade", "fade")
```
Comment (OrsellGit) [inline src/ModFiles/scripts/vscripts/multiplayermod/mapsupport/portal2/sp_a2_bts1.nut]: Clean this up?

File: src/ModFiles/scripts/vscripts/multiplayermod/vars&funcs.nut [hunk 6]
```diff
@@ -2099,4 +2103,38 @@ function FindPlayerByName(name) {
         }
     }
     return best
-}
\ No newline at end of file
+}
+
+function StartCountTransition(player) {
```
Comment (OrsellGit) [inline src/ModFiles/scripts/vscripts/multiplayermod/vars&funcs.nut:2108]: Some statement structures could be optimized here to make it easier to read and run.
For example, instead of surrounding everything, the first if statement can check if the player handle doesn't exist then return. This all can be in one line also.

**Review by OrsellGit — CHANGES_REQUESTED** [review]
There are some questions about what you have done. Also, you can optimize the StartCountTransition function a bit to make it easier to read and run smoother.

File: src/ModFiles/scripts/vscripts/multiplayermod/config.nut (hunk 1 above)
Comment (OrsellGit) [reply src/ModFiles/scripts/vscripts/multiplayermod/config.nut]: Maybe 50 or 75

Comment (Back-Slash-N) [reply src/ModFiles/scripts/vscripts/multiplayermod/config.nut]: changed in newest commit

File: src/ModFiles/scripts/vscripts/multiplayermod/config.nut (hunk 2 above)
Comment (Back-Slash-N) [reply src/ModFiles/scripts/vscripts/multiplayermod/config.nut]: fixed in newest commit
(oops)

File: src/ModFiles/scripts/vscripts/multiplayermod/hooks.nut (hunk 3 above)
Comment (Back-Slash-N) [reply src/ModFiles/scripts/vscripts/multiplayermod/hooks.nut]: removed altogether since P2MM doesn't generally support workshop maps

File: src/ModFiles/scripts/vscripts/multiplayermod/hooks.nut (hunk 4 above)
Comment (Back-Slash-N) [reply src/ModFiles/scripts/vscripts/multiplayermod/hooks.nut:391]: changed in newest commit

File: src/ModFiles/scripts/vscripts/multiplayermod/mapsupport/portal2/sp_a2_bts1.nut (hunk 5 above)
Comment (Back-Slash-N) [reply src/ModFiles/scripts/vscripts/multiplayermod/mapsupport/portal2/sp_a2_bts1.nut]: removed in latest

File: src/ModFiles/scripts/vscripts/multiplayermod/config.nut [hunk 7]
```diff
@@ -46,7 +46,7 @@ Config_UseCountdown <- true
 
 //! Note: Countdown configs are useless if Config_UseCountdown is set to false.
 // Set to the PERCENTAGE of players that you want to reach the end before the transition enables.
-Config_CountdownPercentage <- 100
+Config_CountdownPercentage <- 70
```
Comment (OrsellGit) [inline src/ModFiles/scripts/vscripts/multiplayermod/config.nut]: should be 75, makes more sense and divides more evenly

**Review by OrsellGit — CHANGES_REQUESTED** [review]
Fix plz

**Review by OrsellGit — APPROVED** [review]
Everything is fine, some things are kind of scuffed but in the end, it doesn't even matter, aka we are gonna rewrite this in the plugin later.

**Comment by Back-Slash-N** [comment]
> Everything is fine, some things are kind of scuffed but in the end, it doesn't even matter

caption
